### PR TITLE
Clean up bidding logic

### PR DESCRIPTION
**Pull request by evanakm**

### CODE REVIEW

File: bridgebot/bidding.py [hunk 1]
```diff
@@ -66,21 +66,23 @@ def reset(self):
...
+
     def __increment_player(self):
         self.player = self.player.next_player()
 
     def pass_bid(self):
         if self.bid_index == -1:
             if self.consecutive_passes == 3:  # Passing out
-                self.record[self.player].append(Strains.PASSOUT)
-                self.contract.strain = Strains.NT
+                self.record[self.player].append("PASS")
```
Comment (evanakm) [inline bridgebot/bidding.py]: Not sure the best place for "PASS" to live. Maybe a new `enum`, but that would create some redundancy

File: bridgebot/bidding.py [hunk 2]
```diff
@@ -1,21 +1,123 @@
...
     def __init__(self):
-        self.level = 0
-        self.strain = Strains.NT
+        self.contract = None
         self.doubled = Doubles.NONE
         self.declarer = None
+        self.last_bid_by = None # Redundant, but cleans up edge case logic in Bids.is_sufficient_bid
+
+
+# Compares to a contract, so can't go into enum
```
Comment (evanakm) [inline bridgebot/bidding.py:18]: Not sure whether to put this into `enums.py` or not. It needs the `Contract` class for some of its member functions, so (to me at least) this seems like an intelligent place for it.

File: bridgebot/bidding.py [hunk 3]
```diff
@@ -1,21 +1,123 @@
...
+            Bids.TWO_CLUBS,     Bids.TWO_DIAMONDS,      Bids.TWO_HEARTS,    Bids.TWO_SPADES,    Bids.TWO_NO_TRUMP,
+            Bids.THREE_CLUBS,   Bids.THREE_DIAMONDS,    Bids.THREE_HEARTS,  Bids.THREE_SPADES,  Bids.THREE_NO_TRUMP,
+            Bids.FOUR_CLUBS,    Bids.FOUR_DIAMONDS,     Bids.FOUR_HEARTS,   Bids.FOUR_SPADES,   Bids.FOUR_NO_TRUMP,
+            Bids.FIVE_CLUBS,    Bids.FIVE_DIAMONDS,     Bids.FIVE_HEARTS,   Bids.FIVE_SPADES,   Bids.FIVE_NO_TRUMP,
+            Bids.SIX_CLUBS,     Bids.SIX_DIAMONDS,      Bids.SIX_HEARTS,    Bids.SIX_SPADES,    Bids.SIX_NO_TRUMP,
+            Bids.SEVEN_CLUBS,   Bids.SEVEN_DIAMONDS,    Bids.SEVEN_HEARTS,  Bids.SEVEN_SPADES,  Bids.SEVEN_NO_TRUMP,
+            Bids.PASS,          Bids.DOUBLE,            Bids.REDOUBLE
+        ]
+
+    def is_sufficient_bid(self, by_bidder, current_contract):
```
Comment (evanakm) [inline bridgebot/bidding.py:72]: This encapsulates all the error checking, thus eliminating the need for the functions `pass_bid`, `make_bid`, `double`, and `redouble`. All we need now is one function called `get_new_bid` that calls this function.

File: bridgebot/bidding.py [hunk 4]
```diff
@@ -24,23 +126,50 @@ def __init__(self, first_bidder):
...
 
-    def __ns_first_bid(self, strain):
-        return
-        self.ns_first_bid = {
-            Strains.CLUBS: None,
-            Strains.DIAMONDS: None,
-            Strains.HEARTS: None,
-            Strains.SPADES: None,
-            Strains.NT: None
+        # Not necessary but makes it more human readable
```
Comment (evanakm) [inline bridgebot/bidding.py:129]: Just hard coding this to eliminate the confusion over where to put these dashes.

File: bridgebot/bidding.py [hunk 5]
```diff
@@ -24,23 +126,50 @@ def __init__(self, first_bidder):
...
         }
 
-        self.ew_first_bid = {
-            Strains.CLUBS: None,
-            Strains.DIAMONDS: None,
-            Strains.HEARTS: None,
-            Strains.SPADES: None,
-            Strains.NT: None
-        }
+    def try_to_set_first_bid(self, player, strain):
```
Comment (evanakm) [inline bridgebot/bidding.py:157]: See lines 228-229. It considerably simplifies the logic that was formerly in `make_bid`

File: bridgebot/bidding.py [hunk 6]
```diff
@@ -57,98 +186,55 @@ def reset(self):
         self.player_index = Players.players().index(self.dealer)
         self.last_bidder_index = None
 
-        # Not sure about the following line
```
Comment (evanakm) [inline bridgebot/bidding.py:60]: See comment at line 129

File: bridgebot/bidding.py [hunk 7]
```diff
@@ -57,98 +186,55 @@ def reset(self):
...
 
-    def pass_bid(self):
-        if self.bid_index == -1:
-            if self.consecutive_passes == 3:  # Passing out
-                self.record[self.player].append(Strains.PASSOUT)
-                self.contract.strain = Strains.NT
-                return AuctionStatus.DONE
-            else:  # Passing before an opening bid
-                self.record[self.player].append(Strains.PASSOUT)
+    def get_new_bid(self, new_bid):
```
Comment (evanakm) [inline bridgebot/bidding.py:201]: The crux of refactoring puts everything into this function (and simplifies it)

File: bridgebot/test/test_scoring.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+import sys
+
+sys.path.insert(0,'../bridgebot')
+sys.path.insert(0,'..')
+
+from scoring import get_score_from_result
+from enums import Contracts, Doubles
+
+# Doing this as a black box test
+@pytest.mark.parametrize('contract, doubled, tricks_taken, vulnerability, expected', [
```
Comment (evanakm) [inline bridgebot/test/test_scoring.py:11]: I'm pretty sure this list covers all the different types of cases.

File: bridgebot/bidding.py [hunk 9]
```diff
@@ -24,23 +126,50 @@ def __init__(self, first_bidder):
...
+            self.record[Players.NORTH].append("-")
+        elif dealer == Players.SOUTH:
+            self.record[Players.NORTH].append("-")
+            self.record[Players.EAST].append("-")
+        elif dealer == Players.WEST:
+            self.record[Players.NORTH].append("-")
+            self.record[Players.EAST].append("-")
+            self.record[Players.SOUTH].append("-")
+
+        self.first_bids = {
```
Comment (masonbrothers) [inline bridgebot/bidding.py:140]: This can be determined from the actual bidding record I think. I think this is a duplication of state.

**Review by masonbrothers — APPROVED** [review]

**Pull request merged**